### PR TITLE
[stdlib]remove _PointerFunction

### PR DESCRIPTION
**Pull request by lancep**
The _PointerFunction protocol and related types are kind of weird and unnecessary.

### CODE REVIEW

**Comment by lancep** [comment]
@swift-ci please test

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test linux platform

**Comment by airspeedswift** [comment]
Running full benchmark, some of those regressions are notoriously noisy but you never know.

**Comment by airspeedswift** [comment]
Hmm ok that's a genuine linux failure rather than the spurious one this time.

**Comment by lancep** [comment]
@swift-ci please test

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Comment by lancep** [comment]
@swift-ci please smoke benchmark

**Review by moiseev — COMMENTED** [review]
LGTM

**Comment by stephentyrone** [comment]
This all seems reasonable to me, but then I didn't even know this existed =)

File: stdlib/public/core/ArrayShared.swift [hunk 1]
```diff
@@ -222,17 +198,20 @@ extension _ArrayBufferProtocol {
...
-    Initializer: _PointerFunction,
-    Initializer.Element == Element {
+    _ initializeNewElements: 
+        ((UnsafeMutablePointer<Element>, _ count: Int) -> ())? = nil
+  ) {
 
     _sanityCheck(headCount >= 0)
     _sanityCheck(newCount >= 0)
+    
+    let initializeNewElements = initializeNewElements ?? { ptr, count in
```
Comment (lorentey) [inline stdlib/public/core/ArrayShared.swift]: What if the closure parameter was non-optional, with this as the default value?

**Review by lorentey — APPROVED** [review]
👍 Closures can still be slower sometimes than protocols with a single function requirement (see ObserverClosure vs ObserverForwardingStruct benchmarks), but _PointerFunction seems way over the top

File: stdlib/public/core/ArrayShared.swift [hunk 2]
```diff
@@ -222,17 +198,20 @@ extension _ArrayBufferProtocol {
...
   @inline(never)
   @inlinable // @specializable
-  internal mutating func _arrayOutOfPlaceUpdate<Initializer>(
+  internal mutating func _arrayOutOfPlaceUpdate(
     _ dest: inout _ContiguousArrayBuffer<Element>,
     _ headCount: Int, // Count of initial source elements to copy/move
     _ newCount: Int,  // Number of new elements to insert
-    _ initializeNewElements: Initializer
-  ) where
-    Initializer: _PointerFunction,
```
Comment (lorentey) [inline stdlib/public/core/ArrayShared.swift]: Idea for an alternative design: replace the closure parameter by returning the newly opened uninitialized gap as an `UnsafeMutableBufferPointer`, which the callee would be expected to fill in.

File: stdlib/public/core/ArrayShared.swift (hunk 1 above)
Comment (lancep) [reply stdlib/public/core/ArrayShared.swift]: It makes the function signature pretty gnarly but maybe it's worth it?

File: stdlib/public/core/ArrayShared.swift (hunk 2 above)
Comment (lancep) [reply stdlib/public/core/ArrayShared.swift]: Interesting idea! I think I lean more towards keeping the closure, I like scoping the pointer mutation to as small a scope as possible.

**Comment by lancep** [comment]
@swift-ci please test

**Pull request merged**